File: python/valid_perfect_square.py
class Solution:
    # binary search method
    def isPerfectSquare(self, number: int) -> bool:
        if number == 1:
            return True
        left, right = 0, number // 2
        while left <= right:
            middle = (left + right) // 2
            mid_square = middle * middle
            if mid_square == number:
                return True
            elif mid_square < number:
                left = middle + 1
            else:
                right = middle - 1
        return False

    # Newton's Method
    # def isPerfectSquare(self, number: int) -> bool:
    #     r = number
    #     while r * r > number:
    #         r = (r + number / r) // 2
    #     return r * r == number

File: src/ValidPerfectSquare.java
public class ValidPerfectSquare {
    public static boolean isPerfectSquare(long number) {
        long left = 0, right = number, mid;
        while (left <= right) {
            mid = left + (right - left) / 2;
            if (mid * mid == number) return true;
            else if (mid * mid < number) left = mid + 1;
            else right = mid - 1;
        }
        return false;
    }
}
